relay-server/src/processing/mod.rs

```diff
@@ -0,0 +1,117 @@
+//! A new experimental way to structure Relay's processing pipeline.
+//!
+//! The idea is to slowly move processing steps from the original
+//! [`processor`](crate::services::processor) module into this module.
+//! This is where all the product processing logic should be located or called from.
+//!
+//! The processor service, will then do its actual work using the processing logic defined here.
+
+use relay_config::Config;
+use relay_dynamic_config::GlobalConfig;
+use relay_quotas::RateLimits;
+
+use crate::Envelope;
+use crate::services::processor::ProcessingExtractedMetrics;
+use crate::services::projects::project::ProjectInfo;
+use crate::utils::ManagedEnvelope;
+
+mod counted;
+mod limits;
+pub mod logs;
+mod managed;
+mod utils;
+
+pub use self::counted::*;
+pub use self::limits::*;
+pub use self::managed::*;
+pub use self::utils::*;
+
+/// A processor, for an arbitrary unit of work extracted from an envelope.
+///
+/// The processor takes items from an envelope, to process it.
+/// To fully process an envelope, multiple processor may need to take items from the same envelope.
+///
+/// Event based envelopes should only be handled by a single processor, as the protocol
+/// defines all items in an event based envelope to relate to the envelope.
+pub trait Processor {
+    /// A unit of work, the processor can process.
+    type UnitOfWork: Counted;
+    /// The result after processing a [`Self::UnitOfWork`].
+    type Output: Forward;
+    /// The error returned by [`Self::process`].
+    type Error: std::error::Error;
+
+    /// Extracts a [`Self::UnitOfWork`] from a [`ManagedEnvelope`].
+    ///
+    /// This is infallible, if a processor wants to report an error,
+    /// it should return a [`Self::UnitOfWork`] which later, can produce an error when being
+    fn prepare_envelope(&self, envelope: &mut ManagedEnvelope)
+    -> Option<Managed<Self::UnitOfWork>>;
+
+    /// Processes a [`Self::UnitOfWork`].
+    async fn process(
+        &self,
+        work: Managed<Self::UnitOfWork>,
+        ctx: Context<'_>,
+    ) -> Result<Output<Self::Output>, Rejected<Self::Error>>;
+}
+
+/// Read-only context for processing.
+#[derive(Copy, Clone)]
+pub struct Context<'a> {
```

Comment:
Some processors (e.g. errors, transactions, standalone) also require the `ProjectId`. Is it intended that more fields get added to this struct when we want to port a processor that requires them?